### ISSUE TITLE
C-API for OSResults (orientation and script results) is not a stable ABI

### ISSUE THREAD

**Issue by jbarlow83** [tesseract-ocr/tesseract#424]
The existing C API for `TessBaseAPIDetectOS` requires a C caller to successfully allocate `struct OSResults` which is actually a C++ class. If the C++ compiler decides to append anything to the structure, this will fail. The definition of this struct also depends on `kMaxNumberOfScripts` so is subject to change if (when) Tesseract adds more scripts.

The struct members that a caller is most likely to want are actually in `OSBestResult` whose position may shift.

I suggest deprecating this API and providing one that returns the member elements of `OSBestResult` instead. I think it would be useful to have a C API call to obtain this information rather than parsing the executable's output in `-psm 0`. Here's one possible API.

``` C
BOOL TessBaseAPIDetectOS2(
 TessBaseAPI* handle, 
 int* best_script_id, 
 int* best_orientation_id, 
 float* script_confidence, 
 float* orientation_confidence);
```

Returning `OSBestResults` is also an option, but sometimes it's nice to avoid structs entirely.

I can contribute changes if there's agreement on them.

**Comment by zdenop** [comment]
Thank you for you investigation.
Can you please post it to tesseract-dev forum for discussion, so developers of tesseract wrappers can be aware of change?

**Comment by jbarlow83** [comment]
Posted

**Comment by amitdo** [comment]
The suggestion looks good to me. 

@zdenop, I suggest to give @jbarlow83 a green light to sending a PR.

**Comment by stweil** [comment]
I must admit that I had to look into the code what "OS" means in the context of OCR. Maybe it is possible to find a better name with "Orientation and Script" ("OaS"?).

PS. I just noted that the title of this issue also had the information which I was looking for.

**Comment by zdenop** [comment]
@jbarlow83 will you create PR?

**Comment by jbarlow83** [comment]
Yes, working on it.

**Comment by zdenop** [comment]
fixed with 3a47adcbe13e34
